visualization/app/codeCharta/util/nodeDecorator.ts

```diff
@@ -16,220 +15,210 @@ import { CodeMapHelper } from "./codeMapHelper"
 		map: CodeMapNode,
-		blacklist: BlacklistItem[],
 		metricData: NodeMetricData[],
 		edgeMetricData: EdgeMetricData[],
 		isDeltaState: boolean,
 		attributeTypes: AttributeTypes
 	) {
-		if (map) {
-			const root = hierarchy<CodeMapNode>(map)
-			root.each((node: HierarchyNode<CodeMapNode>) => {
-				const leaves: HierarchyNode<CodeMapNode>[] = node.leaves().filter(x => !x.data.isExcluded)
-				this.decorateNodeWithAggregatedChildrenMetrics(leaves, node, metricData, isDeltaState, attributeTypes)
-				this.decorateNodeWithChildrenSumEdgeMetrics(leaves, node, edgeMetricData, attributeTypes)
-			})
-		}
-		return map
-	}
+		const root = hierarchy<CodeMapNode>(map)
+		root.each((node: HierarchyNode<CodeMapNode>) => {
+			let edgeMetrics;
+			const { data } = node
 
-	private static decorateNodeWithAggregatedChildrenMetrics(
-		leaves: HierarchyNode<CodeMapNode>[],
-		node: HierarchyNode<CodeMapNode>,
-		metricData: NodeMetricData[],
-		isDeltaState: boolean,
-		attributeTypes: AttributeTypes
-	) {
-		metricData.forEach(metric => {
-			if (node.data.children && node.data.children.length > 0) {
-				node.data.attributes[metric.name] = this.aggregateLeafMetric(
-					leaves.map(x => x.data.attributes),
-					metric.name,
+			if (!hasChildren(data)) {
+				return
+			}
+
+			for (const { name } of edgeMetricData) {
+				data.edgeAttributes[name] = { incoming: 0, outgoing: 0 }
+			}
+			const leafAttributes = []
+			const leafDeltas = []
+			// TODO: This provides a huge potential for optimization. The algorithm
+			// should be linear instead of O(n * m) [root.each() * node.leaves()] by
+			// using a depth first algorithm. Each current node tracks all
+			// `attributes`, `deltas` and `edgeAttributes` of the outer nodes
+			// combined. That way there's no need to traverse any node more than once.
+			//
+			// It is also possible to aggregate the stats linear. It's straight
+			// forward for non-median stats (the common case): these are just counted
+			// together. A map should be used to track the different depth.
+			// Calculating the median is more complex, both memory and computational
+			// wise. Instead of calculating the stats together of all outer nodes, an
+			// array of these should be collected (only if required!). New entries are
+			// then added to the array where they fit using array.splice. That way
+			// finding the median is linear instead of O(n log n). It is likely
+			// possible to optimize the latter further but that is probably too much
+			// work.
+			for (const { data } of node.leaves()) {
+				if (data.isExcluded) {
+					continue;
+				}
+				leafAttributes.push(data.attributes)
+				if (data.deltas !== undefined) {
+					leafDeltas.push(data.deltas)
+				}
+				if (!data.edgeAttributes) {
+					continue;
+				}
+				// Optimize for the common case
+				for (const [name, value] of Object.entries(data.edgeAttributes)) {
+					if (!value) {
+						continue;
+					}
```

Comment:
Same here: We set the `edgeAttributes` for each node to `data.edgeAttributes[name] = { incoming: 0, outgoing: 0 }`.
Vale should always be defined.

Reply:
Our unit tests contain entries that have undefined values. Thus, this safeguard should be kept for now.